### ISSUE TITLE
Tagged src does not match pip release set

### ISSUE THREAD

**Issue by mwilsoninsight** [Azure/azure-sdk-for-python#5058]
I'm attempting to package the azure-sdk-for-python and all of its modules for nixpkgs, but I'm having a rough go of it due to version mismatches between pip releases and git tags in your src repository.

For instance, azure-servicemanagement-legacy is offered via pip at version 0.20.6. This is the version you get whether you install the module directly, or as part of the `azure-4.0.0` umbrella module. However, the src code on GitHub that has been tagged as azure_4.0.0 offers the servicemanagement-legacy module at version 0.20.7.

Similarly, azure-nspkg (a dependency for a number of these modules) is offered via pip at version 3.0.2, while the 4.0.0-tagged src code offers it at 2.0.0...

This wouldn't be a bother if I could clone a specific directory, but as it stands, I need to clone the entire repository to build even one package. I can reuse a clone for building multiple packages, but only if the _current versions_ of the packages I want are contained within that commit- otherwise I'm back to cloning the same giant repository for one module. I would re-use the same git clone and just check out different commits, but I'm using an automation tool and to my knowledge that is not yet a feature- instead the only means I know of to change commits is to fetch the entire repo at those specific commits.

**Comment by kurtzeborn** [comment]
Thank you for opening this issue! We are routing it to the appropriate team for follow up.

CC: @lmazuel, @scbedd

**Comment by scbedd** [comment]
So here's what I see:

1. azure-servicemanagement-legacy is checked in at 0.20.7, but is only on pip at version 0.20.6
2. Essentially the issue in a nutshell: the core issue is that the umbrella releases don't happen every time we release an update to any of the packages under the umbrella

**Comment by mwilsoninsight** [comment]
I took another look at the pypi page for `azure` and I think I can now see my confusion.

So if I understand correctly, the version of the umbrella package, `azure-4.0.0`, remains constant even though the patch (and sometimes minor) versions of the contained packages may advance at any given time.
Does that mean the `azure-4.0.0` entry in pip is frequently being re-written and/or could provide different code if downloaded at two separate points in time (if so, why is it constantly v4.0.0)?

If that understanding is true this makes the `azure_4.0.0` tag in the src code not-so-representative of what may currently be available via pip.

Because the contained packages are released much more frequently, is there a single location where one could obtain the src for `azure-latest` (the whole set of up-to-date released packages)?

**Comment by scbedd** [comment]
I will defer to @lmazuel for the absolute correct answer as to rewriting the azure-4.0.0 pip entry, but I will say that I do **not** believe we currently do this. PyPI doesn't allow this that I am aware of. I think we'd need to bump the revision, update the `setup_requires` for whichever packages have updated, and cut a new release for `azure-4.0.X`

>Does that mean the azure-4.0.0 entry in pip is frequently being re-written and/or could provide different code if downloaded at two separate points in time (if so, why is it constantly v4.0.0)?

>If that understanding is true this makes the azure_4.0.0 tag in the src code not-so-representative of what may currently be available via pip.

Yeah, pretty much. The azure_4.0.0 tag is the representation of the state of the repo at the time that the azure-4.0.0 pip package was released. 

>Because the contained packages are released much more frequently, is there a single location where one could obtain the src for azure-latest (the whole set of up-to-date released packages)?

This is a really good question, waiting for @lmazuel to get from pycon to discuss that.

**Comment by mwilsoninsight** [comment]
Fair enough.

> The azure_4.0.0 tag is the representation of the state of the repo at the time that the azure-4.0.0 pip package was released.

That I understand.
However, the state of the repo at `azure_4.0.0` contains `azure-servicemanagement-legacy-0.20.7`.
And pip offers `azure-servicemanagement-legacy` only at `0.20.6` (whether you download the umbrella module or the module itself).

I may be missing something obvious, but the only way I can think of that makes this possible is if `azure-4.0.0` on pip is not actually the src available here, but rather a composition of all of the sub-module releases that are available when a given `azure_X.Y.Z` tag is created. Is this the case?

**Comment by scbedd** [comment]
I think this is confusion stemming from the missed release of `azure-servicemanagement-legacy 0.20.7` 

The latest checked-in source should always match pypi. However, in the case of `azure-servicemanagement-legacy`, 0.20.7 is checked in, but on PyPI it's 0.20.6. This is the source of confusion, and will be remedied. 

[Azure_4.0.0 setup.py](https://github.com/Azure/azure-sdk-for-python/blob/ab01fc1f23462f130c69f46505524b88101023dc/azure/setup.py)

The state of the repo has `azure-servicemanagement-legacy` at 0.20.7, but the actual setup.py version 
specifier for `azure_4.0.0` _is_ ~=0.20.6. See what I mean? 

The world makes sense if `azure-servicemanagement-legacy` version 0.20.7 is released prior to cutting `azure_4.0.0`

Then, `azure_4.0.0` tag represents the state of the repository at the time of the release as we expect it to. Correct? Let's say you manually bump the version specifier on the `azure_4.0.0` `setup.py` to ~=0.20.7. Does your automation run into any other issues with building packages locally from source?

**Comment by mwilsoninsight** [comment]
Ah- indeed, I missed that file. Thanks very much for pointing that out. =]

I assume something similar explains the `azure-nspkg` case as well (I just don't see it mentioned in the `install_requires` there- I'm guessing it's mentioned as a `v3.0.2` dependency in some other `setup.py`).

> The state of the repo has azure-servicemanagement-legacy at 0.20.7, but the actual setup.py version
specifier for azure_4.0.0 is ~=0.20.6. See what I mean?

Yup. Makes perfect sense now that I'm looking at the right file.

> Does your automation run into any other issues with building packages locally from source?

Not really. Hoping some context will clarify:

Currently, I just want one place where I can get the most up-to-date, tested, working, released src (GitHub is nice because the tests aren't stripped and the user-base can run them if they so choose).

Pip is easy for most packages, but the main reason I'm forced to GitHub for any code is because, during ~Feb 2018, some edits were made for wheel 0.31.0 compatibility.
This series of commits makes building from src using wheel possible in nix for _all_ of the modules included in `azure-4.0.0`.
However, the current set of azure modules available in pip omits these commits for ~5 or so packages, making them all fail out with an error:
`error: error in setup.cfg: command 'bdist_wheel' has no such option 'azure_namespace_package'`

I agree it's likely the result of a "missed release" or two along the way, but regardless, the result is having to use multiple commits of the same repo to reliably build each of the contained packages from src (the automation caches, which is great, but, in this case, means multiple copies of the same repo, which is asinine). Is there a way to set up a tag that could represent such a state of the repository (in which each module is checked in at its current, officially-supported version)?

**Comment by lmazuel** [comment]
I see many many questions here, let me try to summarize what I get and let me know if I miss something!

- Yes, azure-servicemanagement-legacy should have been released as 0.20.7. That's on me, I checked a Python 3.7 update a few months before the Python 3.7 was actually released, decided to wait for Py 3.7 to be released and .... forgot. At least, I know no-one using this package with Py 3.7 now :p 
- Tags are meaningful only to the folder they are linked. azure-foo_2.0.0 will just have contract to represent the "azure-foo" folder and nothing else. "azure_4.0.0" represents the "azure" folder and nothing else. Simple as that :). And yes it could mean there is 10 tags to the same commit, if I release a bunch of packages in one shot.
- `error: error in setup.cfg: command 'bdist_wheel' has no such option 'azure_namespace_package'` can be avoided for old packages by downgrading to wheel 0.30.0, or updating to the package where simplified the packaging story (details: https://github.com/Azure/azure-sdk-for-python/wiki/Azure-packaging). If there is still package out there that are not compatible with 0.31.0, feel free to make me a list and I will re-publish a patch version fix

Does that answer all questions? Sorry if I missed some, that's a long issue to catch-up :p
@mwilsoninsight feel free to reach out to me by email as well if necessary (use my github alias and add @microsoft.com ;))
